File: solver.cs
// -*- coding:utf-8 -*-

//
// Pentomino Puzzle Solver with C#
//

using System;
using System.Linq;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;

using Fig = System.Collections.Generic.List<int[]>;

////////////////////////////////////////////////////////

public class Piece
{
    public List<Fig> figs = new List<Fig>();
    public Piece     next;
    public char      id;

    string to_s( Fig fig )
    {
        var xys = fig.Select( xy => string.Format( "({0},{1})", xy[0], xy[1] ));
        return "[" + string.Join( ",",  xys ) + "]";
    }

    string to_s( List<Fig> figs )
    {
        return string.Join( "\n", figs.Select( fig => "    " + to_s( fig ) ) );
    }

    public Piece( char id_ , List<int[]> def, Piece next_ )
    {
        id   = id_;
        next = next_;

        for ( int r_f = 0; r_f < 8; ++r_f ) {     // rotate and flip
            var fig = new Fig();
            foreach( int[] xy in def ) {
                var pt = new int[]{ xy[0], xy[1] };
                for ( int rot = 0; rot < r_f % 4; ++rot ) {      // rotate
                    pt = new int[]{ -pt[1], pt[0] };
                }
                if ( r_f >=4 ) { pt[0] = -pt[0]; }               // flip
                fig.Add( pt );
            }

            fig.Sort( (p,q) =>                                   //sort
                      ( p[1] == q[1] ? (p[0] - q[0]) : (p[1] - q[1]) )
            );

            for ( int j = fig.Count - 1; j >= 0; --j ) {         // normalize
                fig[j] = new int[]{ fig[j][0] - fig[0][0], fig[j][1] - fig[0][1] };
            }

            if ( ! to_s( figs ).Contains( to_s(fig ) ) ) {       //uniq
                this.figs.Add( fig );
            }
        }

        if ( Solver.debug_flg ) {
            Console.WriteLine( "{0}: ({1})", id_, figs.Count );
            Console.WriteLine( "{0}", to_s( figs ) );
        }
    }

}

////////////////////////////////////////////////////////

public class Board
{
    public const char SPACE = ' ';
    public  int  width, height;
    char [,]     cells;

    public Board( int w, int h )
    {
        width  = w;
        height = h;
        cells  = new char [ width, height ];
        for ( int y = 0; y < height; y++ )  {
            for ( int x = 0; x < width; x++ ) {
                cells[ x, y ]  = SPACE;
            }
        }
        if ( width * height == 64 ) {
            Fig hole = new Fig{ new int[] {0,0}, new int[] {0,1},
                                new int[] {1,0}, new int[] {1,1} };
            place( width/2 - 1, height/2 - 1, hole, '@' );
        }
        ELEMS = new string[][]{ elems[0].Split( ',' ), elems[1].Split( ',' ) };
    }

    char at( int x, int y )
    {
        return ( x >= 0 && x < width && y >= 0 && y < height )?
            cells[ x, y ] : '?';
    }


    public bool check( int x, int y, Fig fig )
    {
        return fig.All( pt => at( x + pt[0], y + pt[1] ) == SPACE );
    }


    public void place( int x, int y, Fig fig, char id )
    {
        foreach( var pt in fig ) { cells[ x + pt[0], y + pt[1] ] = id; }
    }


    public int[] find_space( int x, int y )
    {
        while ( cells[ x, y ] != SPACE ) {
            if ( ++x == width ) { x = 0;  ++y; }
        }
        return new int[]{ x, y };
    }

    //         2
    // (-1,-1) | (0,-1)
    //   ---4--+--1----
    // (-1, 0) | (0, 0)
    //         8
    static string[] elems =
    {
        //0      3     5    6    7     9    10   11   12   13   14   15
        "    ,,,+---,,----,+   ,+---,,+---,|   ,+---,+   ,+---,+   ,+---",
        "    ,,,    ,,    ,    ,    ,,|   ,|   ,|   ,|   ,|   ,|   ,|   ",
    };
    static string[][] ELEMS;

    public string render()
    {
        var  lines = new List<string>();

        for ( int y = 0; y <= height; y++ ) {
            for ( int d = 0; d < 2; d++ ) {
                string line = "";
                for ( int x = 0; x <= width; x++ ) {
                    int code =
                        ( ( at( x+0, y+0 ) != at( x+0, y-1 ) )? 1 : 0 ) |
                        ( ( at( x+0, y-1 ) != at( x-1, y-1 ) )? 2 : 0 ) |
                        ( ( at( x-1, y-1 ) != at( x-1, y+0 ) )? 4 : 0 ) |
                        ( ( at( x-1, y+0 ) != at( x+0, y+0 ) )? 8 : 0 );
                    line += ELEMS[d][code];
                }
                lines.Add( line );
            }
        }
        return string.Join( "\n", lines );
    }

}

////////////////////////////////////////////////////////

public class Solver
{
    static string PIECE_DEF_DOC = @"
        +-------+-------+-------+-------+-------+-------+
        |       |   I   |  L    |  N    |       |       |
        |   F F |   I   |  L    |  N    |  P P  | T T T |
        | F F   |   I   |  L    |  N N  |  P P  |   T   |
        |   F   |   I   |  L L  |    N  |  P    |   T   |
        |       |   I   |       |       |       |       |
        +-------+-------+-------+-------+-------+-------+
        |       | V     | W     |   X   |    Y  | Z Z   |
        | U   U | V     | W W   | X X X |  Y Y  |   Z   |
        | U U U | V V V |   W W |   X   |    Y  |   Z Z |
        |       |       |       |       |    Y  |       |
        +-------+-------+-------+-------+-------+-------+
    ";
    static List<int[]> piece_def( char id )
    {
        var def = new List<int[]>();
        int y = 0;
        foreach( var line in PIECE_DEF_DOC.Split( '\n' ) ) {
            for( var x = 0; x < line.Length; x++ ) {
                if ( line[ x ] == id ) { def.Add( new int[]{ x/2, y } ); }
            }
            y++;
        }
        return def;
    }

    public static bool debug_flg = false;

    Piece Unused;
    Board board;
    int solutions = 0;

    Solver( int width, int height )
    {
        Piece pc = null;
        foreach( char id in "FLINPTUVWXYZ".Reverse() ) {
            pc = new Piece( id, piece_def( id ), pc );
        }
        Unused = new Piece( 'F', new List<int[]>(), pc );    // dummy piece
        // limit the symmetry of 'F'
        pc.figs = pc.figs.GetRange( 0, (width == height) ? 1 : 2 );

        board = new Board( width, height);
    }

    void solve( int x, int y )
    {
        if ( Unused.next != null ) {
            var xy  = board.find_space( x, y );
            x = xy[0];  y = xy[1];
            Piece prev = Unused;
            Piece pc;
            while ( (pc = prev.next) != null )
            {
                prev.next = pc.next;
                foreach( Fig fig in pc.figs ) {
                    if ( board.check( x, y, fig ) ) {
                        board.place( x, y, fig, pc.id );
                        solve( x, y );
                        board.place( x, y, fig, Board.SPACE );
                    }
                }
                prev.next = ( prev = pc );
            }
        }
        else {
            solutions ++;
            string curs_up = ( solutions > 1 ) ?
                string.Format( "\x1b[{0}A", board.height * 2 + 2 ) : "";

            Console.WriteLine( curs_up + board.render() + solutions );
        }
    }


    public static void Main( string[] args )
    {
        var width  = 6;
        var height = 10;
        foreach ( string av in args )
        {
            if ( av == "--debug" )  debug_flg = true;

            var m = Regex.Match( av, @"(\d+)\D(\d+)" );
            if (m.Success)
            {
                var w = int.Parse( m.Groups[1].Value );
                var h = int.Parse( m.Groups[2].Value );
                if ( w >= 3 && h >= 3 && ( w * h == 60 | w * h == 64 ) ) {
                    width  = w;
                    height = h;
                }
            }
        }
        var solver = new Solver( width, height );
        solver.solve( 0, 0 );
    }
}

File: solver.vb
' -*- coding:utf-8 -*-

'
' Pentomino Puzzle Solver with Visual Basic
'

Imports System
Imports System.Linq
Imports System.Collections.Generic
Imports System.Text.RegularExpressions

Module Program

  Dim PIECE_DOC =
    "+-------+-------+-------+-------+-------+-------+." &
    "|       |   I   |  L    |  N    |       |       |." &
    "|   F F |   I   |  L    |  N    |  P P  | T T T |." &
    "| F F   |   I   |  L    |  N N  |  P P  |   T   |." &
    "|   F   |   I   |  L L  |    N  |  P    |   T   |." &
    "|       |   I   |       |       |       |       |." &
    "+-------+-------+-------+-------+-------+-------+." &
    "|       | V     | W     |   X   |    Y  | Z Z   |." &
    "| U   U | V     | W W   | X X X |  Y Y  |   Z   |." &
    "| U U U | V V V |   W W |   X   |    Y  |   Z Z |." &
    "|       |       |       |       |    Y  |       |." &
    "+-------+-------+-------+-------+-------+-------+."

' -------------------------------------------------------
  Dim debug_flg = False

  Structure Point
    Public X As Integer
    Public Y As Integer

    Sub New(x As Integer, y As Integer)
        Me.X = x
        Me.Y = y
    End Sub

    Function to_s() As String
      Return String.Format( "({0:D}, {1:D})", X, Y )
    End Function

  End Structure  'Point


  Class CFig
    Inherits List(Of Point )

    Sub New()
    End Sub

    Sub New( src As List(Of Point) )
      MyBase.New(src)
    End Sub

    Function to_s() As String
      Return String.join( ", ", Me.Select( Function( pt ) pt.to_s() ) )
    End Function
  End Class  'CFig


  Function PieceDef( id As Char ) As CFig
    Dim fig As New CFig()
    Dim xy As New Point( 0, 0 )
    For Each c in PIECE_DOC.toCharArray()
      If c = id Then  fig.Add( New Point( xy.X \ 2, xy.Y ) )
      xy.X += 1
      If c = "."c Then xy = New Point( 0, xy.Y + 1 )
    Next
    Return fig
  End FUnction

' -------------------------------------------------------

  Class CPiece
    Public Figs As List(Of CFig )
    Public Id   As Char
    Public Nxt  As CPiece

    Sub New( id_ As Char, org_fig As CFig, nxt_ As CPiece )
      Id   = id_
      Nxt  = nxt_
      Figs = New List(Of CFig )()

      For r_f = 0 to 7
        Dim fig = New CFig()

        For Each pt In org_fig    ' rotate & flip
          For i = 0 To  r_f mod 4                         ' rotate
             pt = New Point( -pt.Y, pt.X )
          Next

          If r_f >= 4 Then pt.X = -pt.X                   ' flip

          fig.Add( pt )
        Next  'pt

        ' sort
        fig.Sort( Function(a, b) If( a.Y = b.Y, a.X - b.X, a.Y - b.Y ) )

        ' normalize
        If fig.Count > 0 Then
          Dim o = fig.Item(0)
          fig = New CFig( fig.ConvertAll( _
                            Function(p) New Point(p.X-o.X, p.Y-o.Y) ))
        End If

        ' uniq
        If Not Figs.Exists( Function(lst) lst.SequenceEqual(fig) ) Then _
          Figs.Add(fig)

      Next  'r_f

      If debug_flg Then
        Console.WriteLine( id & ": ({0:D})", Figs.Count )
        For Each fig In Figs
          Console.WriteLine( "    [ {0} ]", fig.to_s() )
        Next
      End if

    End Sub

  End Class  'CPiece

' -------------------------------------------------------

  Class CBoard
    Public Cells(,) As Char
    Public ReadOnly Width  As Integer
    Public ReadOnly Height As Integer


    Sub New( w As Integer, h As Integer )
      Width  = w
      Height = h
      ReDim Cells( h - 1, w - 1 )

      For y = 0 To h - 1
        For x = 0 To w - 1
          Cells( y, x ) = " "c
        Next
      Next
      If w * h = 64 Then      ' 8x8 or 4x16
        Cells( h \ 2 - 1, w \ 2 - 1 ) = "@"c
        Cells( h \ 2 - 1, w \ 2     ) = "@"c
        Cells( h \ 2    , w \ 2 - 1 ) = "@"c
        Cells( h \ 2    , w \ 2     ) = "@"c
      End If

    End Sub


    Function At( x As Integer, y As Integer ) As Char
      If x >= 0 AndAlso x < Width AndAlso y >= 0 AndAlso y < Height Then
        Return Cells( y, x )
      Else
        Return "?"c
      End If
    End Function


    Function Check( x As Integer, y As Integer, fig As CFig ) As Boolean
      For Each pt In fig
        If At( x + pt.X, y + pt.Y ) <> " "c Then Return False
      Next
      Return True
    End Function


    Sub Place( x As Integer, y As Integer, fig As CFig, id As Char )
      For Each pt In fig
        Cells( y + pt.Y, x + pt.X ) = id
      Next
    End Sub


    Function FindSpace( x As Integer, y As Integer) As Point
      While Cells( y, x ) <> " "c
        x += 1
        If x < Width Then Continue While
        x = 0
        y += 1
      End While
      Return New Point( x, y )
    End Function


    '         2
    ' (-1,-1) | (0,-1)
    '   ---4--+--1----
    ' (-1, 0) | (0, 0)
    '         8
    '          0      3     5    6    7     9    10   11   12   13   14   15
    Const e1="    ,,,+---,,----,+   ,+---,,+---,|   ,+---,+   ,+---,+   ,+---"
    Const e2="    ,,,    ,,    ,    ,    ,,|   ,|   ,|   ,|   ,|   ,|   ,|   "
    Shared ELEMS()() As String = { e1.Split(","c), e2.Split(","c) }

    Function Render() As String
      Dim lines = ""
      For y = 0 To Height
        For d = 0 to 1
          Dim line = ""
          For x = 0 To Width
            Dim code = If( At( x+0, y+0 ) <> At( x+0, y-1 ), 1, 0 ) +
                       If( At( x+0, y-1 ) <> At( x-1, y-1 ), 2, 0 ) +
                       If( At( x-1, y-1 ) <> At( x-1, y+0 ), 4, 0 ) +
                       If( At( x-1, y+0 ) <> At( x+0, y+0 ), 8, 0 )
            line += ELEMS(d)(code)
          Next
          lines += line & vbCrLf
        Next
      Next
      Return lines
    End Function

  End Class  'CBoard

' -------------------------------------------------------

  Class CSolver
    Public Unused As CPiece
    Public board  As CBoard
    Public Solutions = 0
    Public Pieces As New Dictionary(Of Char, CPiece )

    Sub New( w As Integer, h As Integer )
      board = New CBoard( w, h )

      Dim pc As CPiece = Nothing
      For Each id As Char In "FILNPTUVWXYZ".Reverse()
        pc = New CPiece( id, PieceDef( id ), pc )
      Next
      Unused = New CPiece( "!", New CFig(), pc )       ' dummy piece
      ' limit the symetries of "F"
      pc.Figs = pc.Figs.GetRange( 0, If( w = h, 1, 2 ) )
    End Sub


    Sub Solve( x As Integer, y As Integer )
      If Unused.Nxt IsNot Nothing
        Dim xy = board.FindSpace( x, y )
        x = xy.X
        y = xy.Y
        Dim prev = Unused
        While True
          Dim pc = prev.Nxt
          If pc Is Nothing Then Exit While
          prev.Nxt = pc.Nxt
          For Each fig In pc.Figs
            If board.Check( x, y, fig ) Then
              board.Place( x, y, fig, pc.Id )
              Solve( x, y )                    ' call reucrsively
              board.Place( x, y, fig, " "c )
            End If
          Next
          prev.Nxt = pc
          prev     = pc
        End While
      Else
        Solutions += 1
        Dim cursup = If(Solutions > 1, Chr(27)&"["&(board.Height*2+3) &"A", "")
        Console.WriteLine( "{0}{1}{2}", cursup, board.Render, Solutions )
      End If

    End Sub

  End Class  'CSolver

' -------------------------------------------------------

  Public Sub Main( args As String())
    Dim width  = 6
    Dim height = 10
    For Each arg In args
      If arg = "--debug"  Then debug_flg = True
      Dim m = Regex.Match( arg, "^(\d+)\D(\d+)$" )
      If m.Success Then
        Dim w = Integer.Parse( m.Groups(1).Value )
        Dim h = Integer.Parse( m.Groups(2).Value )
        If w >= 3 AndAlso h >= 3 AndAlso ( w * h = 60 OrElse w * h = 64 ) Then
           width  = w
           height = h
        End If
      End If
    Next

    Dim solver = New CSolver( width, height )
    solver.Solve( 0, 0 )
  End Sub

End Module
